### DISCUSSION TITLE
Autonomous Desk

### DISCUSSION THREAD

**Discussion by Tommatheussen** [tjhorner/wifi-desk-controller#16] [Reverse Engineering Help]
I have an Autonomous Desk (https://www.autonomous.ai/en-BE?category_slug=standing-desks&quickview=diy-smart-desk-kit&category=standing-desks) which I wanted to use with the desk controller. Before actually looking into it, I ordered some PCBs from PCBWay (no stock available at the time of ordering, thanks for the PCB designs!).

Now, when they arrived, the controller didn't work, so I started reading up on the possible issues, having seen other desks with pins being changed. After checking with my multimeter and looking at the cables and pins, I came to the conclusion the desk only works with UART, the cable only has 5 pins (5V, GND, Rx, Tx and Sleep)...

So, now my question, I was wondering if you had plans to integrate Serial only desks into the ESPHome config? Or if you would accept PRs that include behavior for these desks?

I'm open to test some things, I hope to be able to map out the pins to the current board layout later today.
I found a couple of repo's the describe the protocol already:
- https://github.com/Stefichen5/AutonomousControl
- https://gist.github.com/developit/f6c7ef3e41a5ac574dbb4af7dd4ec241 (JS implementation of the first link)

**Comment by tjhorner** [comment]
I would definitely love to support serial-only desks in the future. Are the +5V and GND in the same place as the board expects? If so, that would be a simple firmware change.

My time is currently occupied with my dayjob, working on v2 of the board (#14), and I may be moving across the country soon(!) So I probably don't have the time do this myself, but contributions are certainly welcome. Based on the links you provided I think this should be pretty easy to implement entirely in ESPHome yaml config, since it's simple serial messages. You can find an example of something like this in @iMicknl's ESPHome config for LoctekMotion desks: https://github.com/iMicknl/LoctekMotion_IoT/blob/5d9f9c01aa951d8ca24eac3198709b755d330e1b/packages/esphome/flexispot_e6.yaml#L73-L122

Each action is abstracted into a switch entity, which can be used in a script [like I do here](https://github.com/tjhorner/esphome-standing-desk/blob/62d778dc4fe087bff32314a3dc8a1d4fc68101a8/configs/template.yaml#L34-L58).

Let me know if you get this working!

**Reply by Tommatheussen** [reply]
After a bit of measuring and trace testing with my multimeter, I came to the conclusion that the connector used on this desk is actually not an RJ45, but a RJ50 instead. The connector has 10 pins, which essentially wouldn't be a problem, if the cables would have been laid out in such a way that the outer pins are not used, but that's sadly not the case.

I'll need to see if I can order some RJ50 connectors instead and use a prototype board for now to at least get something working

**Reply by Tommatheussen** [reply]
I decided to desolder the wires from the control panel and solder them directly onto the pads of 1 of the RJ45 jacks on the custom PCB.
I now have partial control, I'll attempt to get a fully working ESPHome config up and running this week so I can share more

**Reply by Tommatheussen** [reply]
Something that I now came across as an issue: the UART connection is not designed to have multiple devices listening/sending data. So once my ESP is connected, the control panel doesn't work anymore, removing it does make it work, so the connection does work. I'm thinking if I can use 2 UART busses on the ESP to forward the messages to/from the control separately

**Reply by WilldabeastHA** [reply]
I would be happy to try and contribute here and maybe develop a design specific for this brand. Are all Autonomous desks the same? I have an earlier model I got it about 6-7 years ago. It has the 10 pin connector but only 5 are in use. 
![image](https://user-images.githubusercontent.com/45765518/187433593-dc6c177f-c5f2-4a58-aded-e5539729e8e4.jpg)

**Reply by tjhorner** [reply]
I've reached out to a custom cable manufacturer to get some prototypes of an adapter cable for the Autonomous desks made. Once I get them I'll post another update here and send them to anyone who wants to help test. I don't know what I am going to do with the keypad yet.

You can follow along with the adapter cable saga here: https://github.com/tjhorner/upsy-desky/issues/5

The cables will be backwards compatible with the original WiFi Standing Desk Controller.

**Reply by normanr** [reply]
> Something that I now came across as an issue: the UART connection is not designed to have multiple devices listening/sending data. So once my ESP is connected, the control panel doesn't work anymore, removing it does make it work, so the connection does work. I'm thinking if I can use 2 UART busses on the ESP to forward the messages to/from the control separately

I suspect that the TiMotion devices (which the Autonomous Desks seem to use) might treat the RX/TX lines as shared communication bus, because some of their panels have a RJ50 jack in the back for a bluetooth dongle (TWD1) and it seems (based on continuity tests with a multi-meter) that the wires from the plug are connected directly to the jack. When the control panel is in sleep mode the RX/TX lines float high, which would allow the bluetooth dongle to pull them low to send data (unconfirmed).
